x-pack/apm-server/sampling/eventstorage/storage.go

```diff
@@ -263,7 +263,23 @@ func (rw *ReadWriter) ReadTraceEvents(traceID string, out *modelpb.Batch) error
 	rw.readKeyBuf = append(append(rw.readKeyBuf[:0], traceID...), ':')
 	opts.Prefix = rw.readKeyBuf
 
+	// 1st pass: check whether there exist keys matching the prefix.
+	// Do not prefetch values so that the check is done in-memory.
+	// This is to optimize for cases when it is a miss.
```

Comment:
What is prefetched in case of a miss? Are you implying that badger prefetches stuff that is not relevant to the iterator?

(adding this for future reference)
https://github.com/dgraph-io/badger/blob/f5a7e0a1c83b7db39381e6e528ae23f21a4dd04f/iterator.go#L709

Reply:
> Are you implying that badger prefetches values that are not relevant to the iterator?

Exactly. The line you highlighted, `i.Valid`, `i` is `it.ittr`. `it.ittr.Valid` does not perform prefix checking like `it.Valid`.

Reply:
> What is prefetched in case of a miss?

100 values of irrelevant keys that don't match the supplied key prefix. I _think_ it should be considered a bug in badger v2. Still in process of checking if it is already fixed.

Reply:
While I don't think v4 fixed it, the memory / disk profile changed and I didn't dig into it. https://github.com/carsonip/tbs-badger-playground/tree/badgerv4/prefetch

At the very least, this PR should work on both v2 and v4. It will bring a significant improvement to cases where there's a miss, with little to no impact to happy path (hit).